cns/nodesubnet/ip_fetcher.go

```diff
@@ -0,0 +1,79 @@
+package nodesubnet
+
+import (
+	"context"
+	"log"
+	"time"
+
+	"github.com/Azure/azure-container-networking/nmagent"
+	"github.com/pkg/errors"
+)
+
+// This interface is implemented by the NMAgent Client, and also a mock client for testing
+type ClientInterface interface {
+	GetInterfaceIPInfo(ctx context.Context) (nmagent.Interfaces, error)
+}
+
+type IPFetcher struct {
+	// Node subnet state
+	secondaryIPQueryInterval   time.Duration // Minimum time between secondary IP fetches
+	secondaryIPLastRefreshTime time.Time     // Time of last secondary IP fetch
+
+	ipFectcherClient ClientInterface
+}
+
+func NewIPFetcher(nmaClient ClientInterface, queryInterval time.Duration) *IPFetcher {
+	return &IPFetcher{
+		ipFectcherClient:         nmaClient,
+		secondaryIPQueryInterval: queryInterval,
+	}
+}
+
+// Exposed for testing
+func (c *IPFetcher) SetSecondaryIPQueryInterval(interval time.Duration) {
+	c.secondaryIPQueryInterval = interval
+}
+
+// If secondaryIPQueryInterval has elapsed since the last fetch, fetch secondary IPs
+func (c *IPFetcher) RefreshSecondaryIPsIfNeeded(ctx context.Context) (refreshNeeded bool, ips []string, err error) {
```

Comment:
Generally, if it's not 100% success, it's an error. This includes the entire concept of "retriable" / "non-retriable" errors that we have scattered around (there's good ways of dealing with this if you run into it in the future--ask me). In this case, `refreshNeeded` should be a sentinal error (`ErrRefreshNotNeeded`). That way a caller can just:

```go
ips, err := RefreshSecondaryIPsBlah()
if err != nil && !errors.Is(err, ErrRefreshNotNeeded) {
       // log / alert / etc.
}
```